autorest/codegen/templates/serialization.py.jinja2

```diff
@@ -0,0 +1,2132 @@
+        raise DeserializationError("Cannot deserialize content-type: {}".format(content_type))
+
+    @classmethod
+    def deserialize_from_http_generics(cls, body_bytes, headers):
+        # type: (Optional[Union[AnyStr, IO]], Mapping) -> Any
+        """Deserialize from HTTP response.
+
+        Use bytes and headers to NOT use any requests/aiohttp or whatever
+        specific implementation.
+        Headers will tested for "content-type"
+        """
+        # Try to use content-type from headers if available
+        content_type = None
+        if 'content-type' in headers:
+            content_type = headers['content-type'].split(";")[0].strip().lower()
+        # Ouch, this server did not declare what it sent...
+        # Let's guess it's JSON...
+        # Also, since Autorest was considering that an empty body was a valid JSON,
+        # need that test as well....
+        else:
+            content_type = "application/json"
+
+        if body_bytes:
+            return cls.deserialize_from_text(body_bytes, content_type)
+        return None
+
+class ValidationError(AzureError):
+    """Request parameter validation failed.
+    :param str rule: Validation rule.
+    :param str target: Target value for the rule.
+    :param str value: Value that was invalid.
+    """
+
+    _messages = {
+        "min_length": "must have length greater than {!r}.",
+        "max_length": "must have length less than {!r}.",
+        "minimum_ex": "must be greater than {!r}.",
+        "maximum_ex": "must be less than {!r}.",
+        "minimum": "must be equal to or greater than {!r}.",
+        "maximum": "must be equal to or less than {!r}.",
+        "min_items": "must contain at least {!r} items.",
+        "max_items": "must contain at most {!r} items.",
+        "pattern": "must conform to the following pattern: {!r}.",
+        "unique": "must contain only unique items.",
+        "multiple": "must be a multiple of {!r}.",
+        "required": "can not be None.",
+        "type": "must be of type {!r}"
+    }
+
+    @staticmethod
+    def _format_message(rule, reason, value):
+        if rule == "type" and value.startswith(r"{"):
+            internal_type = value.strip(r"{}")
+            value = "dict[str, {}]".format(internal_type)
+        return reason.format(value)
+
+    def __init__(self, rule, target, value, *args, **kwargs):
+        # type: (str, str, str, str, str) -> None
+        self.rule = rule
+        self.target = target
+        message = "Parameter {!r} ".format(target)
+        reason = self._messages.get(
+            rule, "failed to meet validation requirement.")
+        message += self._format_message(rule, reason, value)
+        super(ValidationError, self).__init__(message, *args, **kwargs)
+
+class DeserializationError(AzureError):
+    """Error raised during response deserialization."""
+    pass
+
+class SerializationError(AzureError):
```

Comment:
once `azure.core` is in, you'll want to throw `azure.core.seriazliation.SerializationError` and `DeserializationError`s instead